src/update_handler.c

```diff
@@ -1921,6 +1940,200 @@ static gboolean img_to_boot_emmc_handler(RaucImage *image, RaucSlot *dest_slot,
 }
 #endif
 
+static gboolean check_if_area_is_clear(const gchar *device, guint64 start, gsize size, gboolean *clear, GError **error)
+{
+	gboolean res = FALSE;
+	g_autofree guchar *read_buf = NULL;
+	gint read_size = 512;
+	gint fd;
+
+	g_return_val_if_fail(device, FALSE);
+	g_return_val_if_fail(clear, FALSE);
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	read_buf = g_malloc0(read_size);
+
+	fd = g_open(device, O_RDONLY);
+	if (fd == -1) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Opening device failed: %s",
+				g_strerror(errno));
+		goto out;
+	}
+
+	if (lseek(fd, start, SEEK_SET) != (off_t) start) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Failed to set file to position %"G_GUINT64_FORMAT ": %s",
+				start, g_strerror(errno));
+		goto out;
+	}
+
+	*clear = TRUE;
+
+	while (size && *clear) {
+		gint read_count = 0;
+
+		if (size < (gsize) read_size)
+			read_size = size;
+
+		read_count = read(fd, read_buf, read_size);
+		if (read_count < 0)
+			goto out;
+
+		for (gint i = 0; i < read_count; i++)
+		{
+			if ((read_buf[i] != 0x00) && (read_buf[i] != 0xFF)) {
+				*clear = FALSE;
+				break;
+			}
+		}
+
+		size -= read_count;
+	}
+
+	res = TRUE;
+
+out:
+	if (fd >= 0)
+		g_close(fd, NULL);
+
+	return res;
+}
+
+static gboolean get_phys_sector_size(const gchar *device, gsize *sector_size, GError **error)
```

Comment:
We don't want to use the physical sector size, as this can be 4kB or 512B depending on the device. Relevant for us is the logical sector size (as read by `get_sectorsize()` in `src/mbr.c`).

If they differ, accesses to a single logical sector are still atomic, but the device may have to perform a read-modify-write cycle, which is what we need here.